Fix normalization of ranges that start at an element node with no children

robertknight opened the pull request:
#73 is caused by a bug in `BrowserRange#normalize()` which is part of the vendor code in `scripts/annotator.js`, where the function failed when normalizing a range whose `startContainer` was an Element node with no children.

This PR fixes the issue in two stages:

1. The first commit imports the latest version of `range.coffee` (plus helpers in `util.coffee` and `xpath.coffee`) from the Annotator v1.2.x source tree directly into our repository, where it is easier to make changes rather than hacking the compiled code. This includes a fix that had already been made in that branch for an issue linked in the commit message.
2. The second commit adds a test case for the issue in #73 and fixes the problem in `BrowserRange#normalize()`

A consequence is that this removes the dependency on Annotator from the anchoring and highlighting code, getting us somewhat closer to a point where we can remove it entirely.

----

* [x] Check licensing issues around the original code. At a minimum, I think we would need to reproduce the MIT license in the files copied from the Annotator 1.2.x codebase.

sean-roberts commented on h/static/scripts/annotator/anchoring/range.coffee:
```diff
@@ -144,7 +144,10 @@ class Range.BrowserRange
     if @startContainer.nodeType is Node.ELEMENT_NODE
       # We are dealing with element nodes
-      r.start = Util.getFirstTextNodeNotBefore @startContainer.childNodes[@startOffset]
+      if @startOffset < @startContainer.childNodes.length
```
Can you add a bit of commentary as to what the problem this check is accounting for?

sean-roberts commented on h/static/scripts/annotator/anchoring/test/html-test.js:
```diff
@@ -98,6 +100,9 @@ var rangeSpecs = [
   [ '/h2[2]',    0,'/p[4]', 0, 'Header Level 2\n\n\n  Mauris lacinia ipsum nulla, id iaculis quam egestas quis.\n\n\n', 'No text node at the end and offset 0'],
+
+  // Additional test cases
+  [ '/gh-73/div', 0, '/gh-73/span/text()', 4, 'Test', 'Range starting at an element node with no children'],
```
I really like the gh-# approach here. Solid 👍

sean-roberts commented:
So far, the code is looking good. I saw that the stuff from #122 was carried over so that's good. Since the bulk of this change is net new files (pulling over), heavily smoke testing is going to be our best bet for this thing. Might want to pull a good bit of the team in to help check a few things.

robertknight commented:
I've done some more local testing and added the license text for the MIT license from the original Annotator codebase to the copied files, in a similar fashion to frame-rpc.js - another module that we earlier copied and modified from elsewhere.

This is ready for review now @sean-roberts
